File: Graph/java/BreadthFirstTraversal.java
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Queue;
import java.util.Set;

public class BreadthFirstTraversal {
  public static List<String> traverse(Map<String, List<String>> graph, String source) {
    List<String> order = new ArrayList<>();
    if (source == null || source.isEmpty()) {
      return order;
    }

    Queue<String> queue = new ArrayDeque<>();
    Set<String> visited = new HashSet<>();

    queue.add(source);
    visited.add(source);

    while (!queue.isEmpty()) {
      String node = queue.poll();
      order.add(node);

      for (String neighbor : graph.getOrDefault(node, List.of())) {
        if (visited.add(neighbor)) {
          queue.add(neighbor);
        }
      }
    }

    return order;
  }
}

File: Graph/csharp/BreadthFirstTraversal.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class BreadthFirstTraversal
{
    public static IList<string> Solve(Dictionary<string, List<string>> graph, string source)
    {
        if (string.IsNullOrEmpty(source))
        {
            return new List<string>();
        }

        var order = new List<string>();
        var queue = new Queue<string>();
        var visited = new HashSet<string>();

        queue.Enqueue(source);
        visited.Add(source);

        while (queue.Count > 0)
        {
            var node = queue.Dequeue();
            order.Add(node);

            if (!graph.TryGetValue(node, out var neighbors))
            {
                continue;
            }

            foreach (var neighbor in neighbors)
            {
                if (visited.Add(neighbor))
                {
                    queue.Enqueue(neighbor);
                }
            }
        }

        return order;
    }
}
